DRILL-6496: Added missing logging statement in VectorUtil.showVectorAccessibleContent(VectorAccessible va, int[] columnWidths)

ilooner opened the pull request.

ilooner commented:
@vvysotskyi please review.

vvysotskyi commented:
Also, after the changes made in DRILL-6438, method `BaseTestQuery.printResult()` do nothing, but it should print the result to stdout.
Could you please fix it?

ilooner commented:
@vvysotskyi This is actually desired behavior since the `VectorUtil.showVectorAccessibleContent` method now uses a logger to print its messages. By default loggers are configure to output to lilith for unit tests, so it will not show up on the console. If you want the log messages to be printed to stdout you need to explicitly enable them with the log fixture:

```
try(LogFixture logFixture = new LogFixture.LogFixtureBuilder()
      .logger("org.apache.drill", Level.OFF) // This is to disable the rest of Drill's debug messages which clog of the screen
      .logger(VectorUtil.class, Level.INFO) // Show results
      .toConsole() // This redirects output to stdout
      .build()) {
  // printResult here
}
```

Then after you are done debugging the test you can remove the log fixture to avoid printing extra messages on travis and jenkins.

vvysotskyi commented:
I disagree with you. A method is called `printResult` but not `logResult` etc.

Regarding log fixture, this method is in `BaseTestQuery`, so I don't think that mixing two testing approaches even for debugging is a good idea.

I don't propose to change the behavior for `VectorUtil.showVectorAccessibleContent()`, may be added a separate method which prints the result to stdout and common logic with `VectorUtil.showVectorAccessibleContent()` may be moved to the separate method.

ilooner commented:
@vvysotskyi The LogFixture is just a utility to adjust log levels, it could be used anywhere and doesn't include any test framework specific code. Also the goal is to remove all usages of **System.out** from the code base as discussed previously https://github.com/apache/drill/pull/1284 , http://mail-archives.apache.org/mod_mbox/drill-dev/201805.mbox/browser and rely completely on the logging framework for output . In order to do that we have to use loggers everywhere we need to output information. I propose to do the following to address your concerns:

1. **VectorUtil.showVectorAccessibleContent()** continues using a logger.
2. **BaseTestQuery.printResult** includes the `try(LogFixture logFixture = new LogFixture.LogFixtureBuilder()...)`boilerplate to output to stdout by default.
3. We add a new method **BaseTestQuery.logResult** which does not turn on output to stdout by default.
4. We replace all usages of **BaseTestQuery.printResult** with **BaseTestQuery.logResult** in the tests.

vvysotskyi commented:
Thanks for the proposal, all these points look good.

arina-ielchiieva commented:
Me might consider the following as well:
1. Rename VectorUtil.showVectorAccessibleContent() -> VectorUtil.logVectorAccessibleContent() 
2. Document usage of LogFixture in dev docs to output the results to stdout.

ilooner-mapr commented:
Thx @vvysotskyi @arina-ielchiieva. I will do the rename of showVectorAccessibleContent and add to the devdocs as well. Also @bitblender just showed me another issue, where he expected the QueryBuilder.printXXX methods to display output as well. So we would have to make a similar adjustment to those methods as well.

arina-ielchiieva commented:
@ilooner maybe can consider allowing stout to ease debugging but fail during project build if left in unit tests?

ilooner commented:
@arina-ielchiieva We can already do that by skipping checkstyle during local development with `-Dcheckstyle.skip`

arina-ielchiieva commented:
@ilooner I meant do we have right now check style for disallowing stout?

ilooner commented:
@arina-ielchiieva We still allow System.out and System.err in our checkstyle rules.

arina-ielchiieva commented:
@ilooner , so let's define what is going to be done in the scope of this PR? Will renaming take place here or in follow up Jira (if second, please create Jira)? Regarding stdout I remember you have created Jira a while ago, so eventually will ban it's usage.

ilooner commented:
@vvysotskyi @arina-ielchiieva I have updated the PR. I found a few more places where methods previously expected to print to stdout, so I ended up changing more methods to make log and print versions of the methods available. Now the scope of the PR is:

- **VectorUtil.showVectorAccessibleContent()** uses a logger and has been renamed to **VectorUtil.logVectorAccessibleContent()**. Additionally the missing log statement was added.
- Print and log variants of methods were created for BaseTestQuery.printResult(), QueryUtils. testRunAndPrint(), QueryBuilder.print(), ClusterTest.runAndLog(), ClientFixture.runQueriesAndLog(). Log methods output to the default log appender. Print methods allow a selected few loggers to output directly to stdout allowing results to be printed to the console for debugging. Log method variants can be called from tests and left in test code. Print method variants should only be used for debugging tests and should not be committed.
- LoggingResultsListener was created in addition to the PrintingResultsListener in order to enable the creation of log and print variants of all methods that output results.
- PrintingUtils.java was added in order to encapsulate the code for turning on logging to stdout, so the same boilerplate is not repeated everywhere.
- Added some more documentation for logging.

Disallowing System.out will be done in this Jira https://issues.apache.org/jira/browse/DRILL-6464

Please let me know if you have anymore comments.

arina-ielchiieva commented on contrib/format-maprdb/src/test/java/com/mapr/drill/maprdb/tests/json/BaseJsonTest.java:
```diff
@@ -59,7 +59,7 @@ protected void runSQLAndVerifyCount(String sql, int expectedRowCount) throws Exc
   }
 
   private void printResultAndVerifyRowCount(List<QueryDataBatch> results, int expectedRowCount) throws SchemaChangeException {
```
I guess we can rename this one to log as well.

ilooner replied:
Fixed.

arina-ielchiieva commented on exec/java-exec/src/test/java/org/apache/drill/test/QueryTestUtil.java:
```diff
@@ -47,9 +46,6 @@
  */
 public class QueryTestUtil {
-
-  public static final String TEST_QUERY_PRINTING_SILENT = "drill.test.query.printing.silent";
```
If we are going to remove this property, I think it should be also removed from `pom.xml` file as well. Please also check for references in other places.

ilooner replied:
Grepped for it and removed references in pom.xml and Testing.md

arina-ielchiieva commented on exec/java-exec/src/test/java/org/apache/drill/test/QueryTestUtil.java:
```diff
@@ -100,36 +96,125 @@ public static String normalizeQuery(final String query) {
+   * @param drillClient drill client to use
+   * @param queryString the query string
+   * @param print True to output results to stdout. False to log results.
+   * @throws Exception
```
Please add description to the Exception here and in the methods below, otherwise IDE shows it as warning.

ilooner replied:
Fixed.

arina-ielchiieva commented on exec/java-exec/src/test/java/org/apache/drill/test/QueryTestUtil.java:
```diff
@@ -100,36 +96,125 @@ public static String normalizeQuery(final String query) {
+  public static void testRunAndOutput(final DrillClient drillClient,
+                                      final String queryString,
+                                      final boolean print) throws Exception{
+    final String query = normalizeQuery(queryString);
```
Looks query extraction logic is similar here and in `testRunAndLog`, `testRunAndPrint`, can be extracted in private method.

ilooner replied:
Fixed.

arina-ielchiieva commented on contrib/storage-kafka/src/test/java/org/apache/drill/exec/store/kafka/KafkaTestBase.java:
```diff
@@ -71,7 +71,7 @@ public void runKafkaSQLVerifyCount(String sql, int expectedRowCount) throws Exce
 
   public void printResultAndVerifyRowCount(List<QueryDataBatch> results, int expectedRowCount)
```
log

ilooner replied:
Fixed.

arina-ielchiieva commented on contrib/storage-mongo/src/test/java/org/apache/drill/exec/store/mongo/MongoTestBase.java:
```diff
@@ -69,7 +69,7 @@ public void runMongoSQLVerifyCount(String sql, int expectedRowCount)
 
   public void printResultAndVerifyRowCount(List<QueryDataBatch> results,
```
log

ilooner replied:
Fixed.

arina-ielchiieva commented on contrib/storage-hbase/src/test/java/org/apache/drill/hbase/BaseHBaseTest.java:
```diff
@@ -93,7 +93,7 @@ protected void runHBaseSQLVerifyCount(String sql, int expectedRowCount) throws E
   }
 
   private void printResultAndVerifyRowCount(List<QueryDataBatch> results, int expectedRowCount) throws SchemaChangeException {
```
log

ilooner replied:
Fixed

arina-ielchiieva commented on exec/java-exec/src/test/java/org/apache/drill/test/PrintingUtils.java:
```diff
@@ -0,0 +1,84 @@
+   * @param <T> The return type of the lambda function.
+   * @param <E> The type of exception thrown by the lambda function.
+   */
+  @FunctionalInterface
```
Maybe move it to `org.apache.drill.exec.util` class? It maybe used in other classes in future.

ilooner replied:
done

ilooner commented:
@arina-ielchiieva Thanks for catching mistakes, I have applied the review comments.

arina-ielchiieva commented:
Looks like there are some compilation errors:
```
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/travis/build/apache/drill/exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/lateraljoin/TestE2EUnnestAndLateral.java:[396,6] error: cannot find symbol
[ERROR]   symbol:   method test(String)
  location: class TestE2EUnnestAndLateral
/home/travis/build/apache/drill/exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/lateraljoin/TestE2EUnnestAndLateral.java:[432,6] error: cannot find symbol
```
The changes look good though, thanks for making changes after code review, Putting +1 here, can be merged when compilation errors are addressed.

sohami commented:
@arina-ielchiieva / @ilooner 
Removing ready-to-commit tag until compilation issue is fixed. Please add it once compilation issue is fixed.

arina-ielchiieva commented:
@ilooner could you please fix compilation errors?

ilooner commented:
@arina-ielchiieva Compilation issue is fixed, missed renaming two method calls. Adding back ready to commit tag.
